File: mobile-rn/src/services/authService.ts
import { secureStore } from './secureStore';
import { SUPABASE_URL, SUPABASE_ANON_KEY, supabaseConfigured } from '../config/env';

/**
 * App auth (multi-user backend, iOS plan §3.4.1), ported from the Swift
 * `AuthService`. It authenticates **directly against Supabase GoTrue's REST API**
 * (`POST /auth/v1/token`) with `fetch` — no `@supabase/supabase-js` dependency
 * required for the token flow (that lib is an option later for realtime/storage).
 *
 * The access token is injected as `Authorization: Bearer <token>` on every
 * `apiClient` / `sseClient` request and refreshed on a 401. Tokens persist in
 * `expo-secure-store` (Keychain / Keystore).
 *
 * The `LoginScreen` (driven by the auth gate in `RootNavigator`) calls `signIn`;
 * `SettingsScreen` calls `signOut`. When signed out, `getAuthHeader()` returns
 * null: the open legacy `/api/*` endpoints still work, but the gated `/api/v2/*`
 * endpoints (e.g. `POST /api/v2/runs`, used for run creation) return 401.
 */

const ACCESS_KEY = 'auth.accessToken';
const REFRESH_KEY = 'auth.refreshToken';
const USER_ID_KEY = 'auth.userId';
const EMAIL_KEY = 'auth.email';

export interface AuthUser {
  id: string;
  email: string | null;
}

interface GoTrueSession {
  access_token: string;
  refresh_token: string;
  user?: { id?: string; email?: string };
}

/**
 * Result of `signUp`. GoTrue's `/signup` returns either a full session (email
 * confirmation OFF) or a user object with no session (confirmation required).
 */
export type SignUpResult =
  | { status: 'session' }
  | { status: 'confirmation_required'; email: string };

type Listener = () => void;

class AuthService {
  private accessToken: string | null = null;
  private refreshToken: string | null = null;
  private user: AuthUser | null = null;
  private loaded = false;
  private refreshInFlight: Promise<boolean> | null = null;
  private listeners = new Set<Listener>();

  /** Whether Supabase sign-in is configured (URL + anon key present). */
  get isConfigured(): boolean {
    return supabaseConfigured;
  }

  get isSignedIn(): boolean {
    return this.accessToken != null;
  }

  /** Whether persisted tokens have been loaded from SecureStore yet. */
  get isHydrated(): boolean {
    return this.loaded;
  }

  get currentUser(): AuthUser | null {
    return this.user;
  }

  subscribe(listener: Listener): () => void {
    this.listeners.add(listener);
    return () => this.listeners.delete(listener);
  }

  private notify() {
    for (const l of this.listeners) l();
  }

  /** Load persisted tokens once on app start. */
  async hydrate(): Promise<void> {
    if (this.loaded) return;
    this.accessToken = await secureStore.get(ACCESS_KEY);
    this.refreshToken = await secureStore.get(REFRESH_KEY);
    const id = await secureStore.get(USER_ID_KEY);
    const email = await secureStore.get(EMAIL_KEY);
    this.user = id ? { id, email } : null;
    this.loaded = true;
    this.notify();
  }

  /** Header injected on every authenticated request (or null when signed out). */
  async getAuthHeader(): Promise<string | null> {
    if (!this.loaded) await this.hydrate();
    return this.accessToken ? `Bearer ${this.accessToken}` : null;
  }

  /**
   * For dev/testing without a Supabase project: inject a pre-minted bearer token
   * directly (e.g. a `MOBILE_AUTH_MODE=dev` HMAC token). Persisted like a real
   * session so `/api/v2/*` calls authenticate.
   */
  async setDevToken(token: string): Promise<void> {
    this.accessToken = token;
    await secureStore.set(ACCESS_KEY, token);
    this.notify();
  }

  /** `POST /auth/v1/token?grant_type=password`. */
  async signIn(email: string, password: string): Promise<void> {
    const session = await this.requestToken('password', { email, password });
    await this.persist(session);
  }

  /**
   * `POST /auth/v1/signup`. Creates a new account. When email confirmation is
   * disabled GoTrue returns a full session (we persist it so the auth gate
   * advances); when confirmation is required it returns a user with no session
   * (we report `confirmation_required` so the UI can prompt the user to check
   * their inbox). Additive: does not touch the sign-in flow.
   */
  async signUp(email: string, password: string): Promise<SignUpResult> {
    if (!this.isConfigured) {
      throw new Error("Sign-up isn't configured. Set EXPO_PUBLIC_SUPABASE_URL / _ANON_KEY.");
    }
    const url = `${SUPABASE_URL.replace(/\/$/, '')}/auth/v1/signup`;
    const res = await fetch(url, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        Accept: 'application/json',
        apikey: SUPABASE_ANON_KEY,
      },
      body: JSON.stringify({ email, password }),
    });
    if (!res.ok) {
      let message = `Sign-up failed (HTTP ${res.status}).`;
      try {
        const data = (await res.json()) as Record<string, unknown>;
        const m = data.error_description ?? data.msg ?? data.message;
        if (typeof m === 'string' && m) message = m;
      } catch {
        // ignore parse failure
      }
      throw new Error(message);
    }
    const data = (await res.json()) as GoTrueSession;
    if (data.access_token) {
      await this.persist(data);
      return { status: 'session' };
    }
    return { status: 'confirmation_required', email };
  }

  /**
   * `POST /auth/v1/recover`. Triggers a password-reset email through Supabase
   * GoTrue (the same endpoint `supabase-js`'s `auth.resetPasswordForEmail`
   * calls). We intentionally treat the response as opaque success/failure: per
   * GoTrue's user-enumeration policy a 200 response does not confirm that the
   * email is registered, so the UI should always show a generic
   * "if an account exists…" confirmation. `redirectTo` is left to the project's
   * default Site URL — completing the reset (deep link → new password screen)
   * is a separate, deferred flow.
   */
  async resetPassword(email: string): Promise<void> {
    if (!this.isConfigured) {
      throw new Error("Password reset isn't configured. Set EXPO_PUBLIC_SUPABASE_URL / _ANON_KEY.");
    }
    const url = `${SUPABASE_URL.replace(/\/$/, '')}/auth/v1/recover`;
    const res = await fetch(url, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        Accept: 'application/json',
        apikey: SUPABASE_ANON_KEY,
      },
      body: JSON.stringify({ email }),
    });
    if (!res.ok) {
      let message = `Password reset failed (HTTP ${res.status}).`;
      try {
        const data = (await res.json()) as Record<string, unknown>;
        const m = data.error_description ?? data.msg ?? data.message;
        if (typeof m === 'string' && m) message = m;
      } catch {
        // ignore parse failure
      }
      throw new Error(message);
    }
  }

  /** Exchange the stored refresh token for a fresh session (called on a 401). */
  async refresh(): Promise<void> {
    if (!this.refreshToken) throw new Error('No refresh token; please sign in again.');
    const session = await this.requestToken('refresh_token', {
      refresh_token: this.refreshToken,
    });
    await this.persist(session);
  }

  /**
   * Best-effort, single-flight refresh used by `apiClient` / `sseClient` on a
   * 401. Returns true if a fresh access token is now stored. Concurrent callers
   * share one in-flight refresh.
   */
  async attemptRefresh(): Promise<boolean> {
    if (this.refreshInFlight) return this.refreshInFlight;
    const task = (async () => {
      try {
        await this.refresh();
        return true;
      } catch {
        return false;
      } finally {
        this.refreshInFlight = null;
      }
    })();
    this.refreshInFlight = task;
    return task;
  }

  async signOut(): Promise<void> {
    const token = this.accessToken;
    this.accessToken = null;
    this.refreshToken = null;
    this.user = null;
    await Promise.all([
      secureStore.delete(ACCESS_KEY),
      secureStore.delete(REFRESH_KEY),
      secureStore.delete(USER_ID_KEY),
      secureStore.delete(EMAIL_KEY),
    ]);
    this.notify();
    // Best-effort server-side revocation; failures are ignored.
    if (token && this.isConfigured) {
      void fetch(`${SUPABASE_URL.replace(/\/$/, '')}/auth/v1/logout`, {
        method: 'POST',
        headers: { apikey: SUPABASE_ANON_KEY, Authorization: `Bearer ${token}` },
      }).catch(() => undefined);
    }
  }

  private async requestToken(
    grantType: 'password' | 'refresh_token',
    body: Record<string, string>,
  ): Promise<GoTrueSession> {
    if (!this.isConfigured) {
      throw new Error("Sign-in isn't configured. Set EXPO_PUBLIC_SUPABASE_URL / _ANON_KEY.");
    }
    const url = `${SUPABASE_URL.replace(/\/$/, '')}/auth/v1/token?grant_type=${grantType}`;
    const res = await fetch(url, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        Accept: 'application/json',
        apikey: SUPABASE_ANON_KEY,
      },
      body: JSON.stringify(body),
    });
    if (!res.ok) {
      let message = `Sign-in failed (HTTP ${res.status}).`;
      try {
        const data = (await res.json()) as Record<string, unknown>;
        const m = data.error_description ?? data.msg ?? data.message;
        if (typeof m === 'string' && m) message = m;
      } catch {
        // ignore parse failure
      }
      throw new Error(message);
    }
    return (await res.json()) as GoTrueSession;
  }

  private async persist(session: GoTrueSession): Promise<void> {
    this.accessToken = session.access_token;
    this.refreshToken = session.refresh_token;
    this.user = session.user?.id
      ? { id: session.user.id, email: session.user.email ?? null }
      : this.user;
    await Promise.all([
      secureStore.set(ACCESS_KEY, session.access_token),
      secureStore.set(REFRESH_KEY, session.refresh_token),
      this.user ? secureStore.set(USER_ID_KEY, this.user.id) : Promise.resolve(),
      this.user?.email ? secureStore.set(EMAIL_KEY, this.user.email) : Promise.resolve(),
    ]);
    this.notify();
  }
}

export const authService = new AuthService();

File: ios/StockAgents/Services/AuthService.swift
import Foundation

/// Real Supabase Auth (GoTrue) sign-in for the multi-user backend (plan §3.4.1).
///
/// The app authenticates **directly against Supabase GoTrue's REST API** with
/// `URLSession` (zero third-party deps — the `supabase-swift` SPM package is
/// intentionally NOT added). A successful sign-in yields a normal Supabase user
/// access token + refresh token; the access token is injected as
/// `Authorization: Bearer <token>` on every `APIClient` / `RunStreamClient`
/// request, and the backend (`web/mobile/auth.py`, `MOBILE_AUTH_MODE=supabase`)
/// validates it (signature + exp + aud + iss).
///
/// Tokens are persisted in the **Keychain** (`KeychainStore`,
/// `kSecAttrAccessibleWhenUnlockedThisDeviceOnly`). The bearer header wiring in
/// the networking layer is unchanged; this type just fills in the real sign-in,
/// refresh, and sign-out flows and exposes an observable `state` the UI drives.
@MainActor
final class AuthService: ObservableObject {
    private static let tokenKey = "auth.accessToken"
    private static let refreshKey = "auth.refreshToken"
    private static let expiryKey = "auth.expiresAt"
    private static let userIDKey = "auth.userID"
    private static let emailKey = "auth.email"

    private let keychain: KeychainStore
    private let config: SupabaseConfig?
    private let session: URLSession
    private let decoder = JSONDecoder()
    private let encoder = JSONEncoder()

    /// Observable sign-in state for the UI (gate login vs. main app).
    @Published private(set) var state: AuthState

    /// Coalesces concurrent refreshes (e.g. several 401s in flight at once) so
    /// only one refresh-token exchange runs and all callers await its result.
    private var refreshTask: Task<Bool, Never>?

    init(keychain: KeychainStore, config: SupabaseConfig? = SupabaseConfig.resolve(), session: URLSession = .shared) {
        self.keychain = keychain
        self.config = config
        self.session = session
        self.state = AuthService.restoreState(from: keychain)
    }

    /// Whether a Supabase project URL + anon key are configured. When false the
    /// login screen shows setup guidance instead of a doomed sign-in attempt.
    var isConfigured: Bool { config != nil }

    var isSignedIn: Bool {
        if case .signedIn = state { return true }
        return false
    }

    // MARK: - Token accessors (read by the networking layer)

    /// Current bearer token, if any.
    var accessToken: String? {
        keychain.get(AuthService.tokenKey)
    }

    /// Header value injected on every authenticated request, or nil when signed
    /// out (the legacy/open endpoints simply ignore its absence).
    var authorizationHeader: String? {
        accessToken.map { "Bearer \($0)" }
    }

    // MARK: - Sign in

    /// `POST {SUPABASE_URL}/auth/v1/token?grant_type=password` with the anon key
    /// as the `apikey` header. Persists the resulting session and flips `state`
    /// to `.signedIn`.
    func signIn(email: String, password: String) async throws {
        guard let config else { throw AuthError.notConfigured }
        let body = try encoder.encode(["email": email, "password": password])
        let session = try await requestToken(grantType: "password", body: body, config: config)
        persist(session)
        state = .signedIn(session.authUser)
    }

    // MARK: - Refresh

    /// Exchange the stored refresh token for a fresh session (called on a 401).
    /// Throws on failure so callers can distinguish a recoverable refresh from a
    /// hard sign-out. Most callers should use `attemptRefresh()` instead.
    func refresh() async throws {
        guard let config else { throw AuthError.notConfigured }
        guard let refreshToken = keychain.get(AuthService.refreshKey) else {
            throw AuthError.noRefreshToken
        }
        let body = try encoder.encode(["refresh_token": refreshToken])
        let session = try await requestToken(grantType: "refresh_token", body: body, config: config)
        persist(session)
        state = .signedIn(session.authUser)
    }

    /// Best-effort, coalesced refresh used by `APIClient` on a 401. Returns
    /// `true` if a fresh access token is now stored. Concurrent callers share a
    /// single in-flight refresh.
    func attemptRefresh() async -> Bool {
        if let refreshTask { return await refreshTask.value }
        let task = Task { () -> Bool in
            do {
                try await self.refresh()
                return true
            } catch {
                return false
            }
        }
        refreshTask = task
        let result = await task.value
        refreshTask = nil
        return result
    }

    // MARK: - Sign out

    /// Clear the local session and flip `state` to `.signedOut`. Also makes a
    /// best-effort GoTrue `logout` call to revoke the refresh token server-side
    /// (failures are ignored — local sign-out always succeeds).
    func signOut() {
        let token = accessToken
        clearTokens()
        state = .signedOut
        if let config, let token {
            Task { await revokeRemotely(token: token, config: config) }
        }
    }

    // MARK: - Persistence

    private func persist(_ session: SupabaseSession) {
        keychain.set(session.accessToken, for: AuthService.tokenKey)
        keychain.set(session.refreshToken, for: AuthService.refreshKey)
        keychain.set(String(session.expiryEpoch), for: AuthService.expiryKey)
        if let user = session.user {
            keychain.set(user.id, for: AuthService.userIDKey)
            if let email = user.email {
                keychain.set(email, for: AuthService.emailKey)
            }
        }
    }

    private func clearTokens() {
        keychain.delete(AuthService.tokenKey)
        keychain.delete(AuthService.refreshKey)
        keychain.delete(AuthService.expiryKey)
        keychain.delete(AuthService.userIDKey)
        keychain.delete(AuthService.emailKey)
    }

    /// Restore sign-in state on launch. A stored access *or* refresh token is
    /// treated as signed-in: an expired access token is recovered by the
    /// 401→refresh path on the first authenticated call.
    private static func restoreState(from keychain: KeychainStore) -> AuthState {
        let hasAccess = keychain.get(tokenKey) != nil
        let hasRefresh = keychain.get(refreshKey) != nil
        guard hasAccess || hasRefresh else { return .signedOut }
        let id = keychain.get(userIDKey) ?? ""
        return .signedIn(AuthUser(id: id, email: keychain.get(emailKey)))
    }

    // MARK: - GoTrue REST

    private func requestToken(grantType: String, body: Data, config: SupabaseConfig) async throws -> SupabaseSession {
        var components = URLComponents(url: config.authBaseURL.appendingPathComponent("token"), resolvingAgainstBaseURL: false)
        components?.queryItems = [URLQueryItem(name: "grant_type", value: grantType)]
        guard let url = components?.url else { throw AuthError.notConfigured }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.timeoutInterval = 30
        request.httpBody = body
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        request.setValue(config.anonKey, forHTTPHeaderField: "apikey")

        let data: Data
        let response: URLResponse
        do {
            (data, response) = try await session.data(for: request)
        } catch {
            throw AuthError.transport(error.localizedDescription)
        }
        guard let http = response as? HTTPURLResponse else {
            throw AuthError.transport("No HTTP response from Supabase.")
        }
        guard (200...299).contains(http.statusCode) else {
            throw AuthError.server(status: http.statusCode, message: AuthService.errorMessage(from: data))
        }
        do {
            return try decoder.decode(SupabaseSession.self, from: data)
        } catch {
            throw AuthError.decoding(String(describing: error))
        }
    }

    private func revokeRemotely(token: String, config: SupabaseConfig) async {
        var request = URLRequest(url: config.authBaseURL.appendingPathComponent("logout"))
        request.httpMethod = "POST"
        request.timeoutInterval = 15
        request.setValue(config.anonKey, forHTTPHeaderField: "apikey")
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        _ = try? await session.data(for: request)
    }

    private static func errorMessage(from data: Data) -> String? {
        guard let body = try? JSONDecoder().decode(SupabaseAuthErrorBody.self, from: data) else { return nil }
        return body.displayMessage
    }

    // MARK: - Errors

    enum AuthError: LocalizedError, Equatable {
        case notConfigured
        case noRefreshToken
        case server(status: Int, message: String?)
        case transport(String)
        case decoding(String)

        var errorDescription: String? {
            switch self {
            case .notConfigured:
                return "Sign-in isn't configured. Set the Supabase URL and anon key (see Settings / ios README)."
            case .noRefreshToken:
                return "Your session expired. Please sign in again."
            case .server(let status, let message):
                if let message, !message.isEmpty { return message }
                if status == 400 || status == 401 { return "Invalid email or password." }
                return "Sign-in failed (HTTP \(status))."
            case .transport(let message):
                return "Network error: \(message)"
            case .decoding(let message):
                return "Could not read the sign-in response: \(message)"
            }
        }
    }
}
